### PR TITLE
add ethereum-vm

### PR DESCRIPTION
**Pull request by diasdavid**
@wanderer @kumavis ~~could you give me a hint, I've added the ethereum-vm to this ethereum-libp2p node, but when running the blockchain I'm getting an error and I'm not being able to identify where did I screw up. Error:~~

```
  1) ethereum-vm run through the blockchain:
     TypeError: this.stateManager.blockchain.iterator is not a function
      at VM.module.exports [as runBlockchain] (node_modules/ethereumjs-vm/lib/runBlockchain.js:20:32)
      at Context.it (test/vm.spec.js:51:12)
```

~~To test it, clone the repo, then checkout this branch, npm install and npm test.~~

**Update:** Nevermind, it is just the ethereumjs-vm version on npm that is out of date.

Also, can you help me break the test-data into pieces? From what I understand, there is always a pre-setup to load a bunch of accounts and then the blocks represent the transactions that happened on those accounts, how can I expose the state of those accounts and make it visual? I'm not totally clear of what is going on that blockchain.

I'm starting to build a WebUI (in the examples folder) and my goal is to:

a) be able to show the vm running the blockchain (an artificial one)
b) boot up new browser nodes and see them fetching the blocks from the first node that has this new, shortlived blockchain, and validating it as well
c) then for the final trick, refresh the page and start fetching the blocks from the ethereum network.

### CODE REVIEW

**Comment by wanderer** [comment]
> Also, can you help me break the test-data into pieces?

what exactly do you need? the descriptions of the different fields in the test data?

> From what I understand, there is always a pre-setup to load a bunch of accounts and then the blocks represent the transactions that happened on those accounts

yep!

>  how can I expose the state of those accounts and make it visual?

each account is [balance, nonce, codeHash, stateRoot]

you should be able to dump the entire state with

[`state.createReadStream`](https://github.com/ethereumjs/merkle-patricia-tree/blob/master/docs/index.md#createreadstream)
```
const readStream = state.createReadStream()
readStream.on('data', (rawAaccount) =>  {
   const account = new Account(rawAccount)
   account.balance
   account.nonce
   ....
})
```

is this the data you want? im not sure how to make visual

File: test/data/vm-data.json [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+      "nonce": "0x00",
+      "storage": {}
+    },
+    "aaaf5374fce5edbc8e2a8697c15331677e6ebf0b": {
+      "balance": "0x02540be400",
+      "code": "0x73a94f5374fce5edbc8e2a8697c15331677e6ebf0bff",
+      "nonce": "0x00",
+      "storage": {}
+    }
+  }
```
Comment (diasdavid) [inline test/data/vm-data.json]: note to self: These are the two initial accounts.

File: test/data/vm-data.json [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+      "nonce": "0x07",
+      "storage": {}
+    },
+    "bbbf5374fce5edbc8e2a8697c15331677e6ebf0b": {
+      "balance": "0x3c",
+      "code": "0x",
+      "nonce": "0x00",
+      "storage": {}
+    }
+  },
```
Comment (diasdavid) [inline test/data/vm-data.json]: note to self: This is the final state of the accounts. One of the operations creates a new account.

File: test/data/vm-data.json [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+  "genesisRLP": "0xf901fcf901f7a00000000000000000000000000000000000000000000000000000000000000000a01dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347948888f1f195afa192cfee860698584c030f4c9db1a0af81e09f8c46ca322193edfda764fa7e88e81923f802f1d325ec0b0308ac2cd0a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421b9010000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000830200008083023e38808454c98c8142a056e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421880102030405060708c0c0",
+  "lastblockhash": "0xf53f268d23a71e85c7d6d83a9504298712b84c1a2ba220441c86eeda0bf0b6e3",
```
Comment (diasdavid) [inline test/data/vm-data.json]: note to self: final step after the operations

File: test/vm-1st-1000.spec.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    console.log(decoded)
+    /*
+    const genesisBlock = new Block()
+
+    genesisBlock.header = new BlockHeader(
+                            testData.genesisBlockHeader
+                          )
+    eth.vm._blockchain.putGenesis(genesisBlock, done)
+    */
+  })
```
Comment (diasdavid) [inline test/vm-1st-1000.spec.js]: @wanderer, got the 1st thousand blocks from the blockchain, now I want to run them on the VM. The blocks are rlp encoded on the `first-1000-blocks.json`. 

With regards to https://github.com/ethereumjs/ethereumjs-vm/pull/86#pullrequestreview-162748, what magic do I need to do?

Also, why do we only need to store the BlockHeader for the genesis, would putting the whole block 0 be an issue?

**Comment by diasdavid** [comment]
Note to self: Main chain initial state: https://github.com/ethereumjs/common/blob/master/genesisState.json
Thanks for pointing this to me @wanderer :)

File: test/vm-with-real-chain-1st-1000.spec.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+
+        console.log('->', block.difficulty)
+
+        eth.vm._blockchain.putBlock(block, (err) => {
+          expect(err).to.not.exist
+          cb
+        })
+      } catch (err) { cb(err) }
+    }
+  })
```
Comment (diasdavid) [inline test/vm-with-real-chain-1st-1000.spec.js]: @wanderer this is the test that is failing, it tries to process the first thousand blocks of the real ethereum chain, but fails on the first one with:

```
  1) process 1st 1000 blocks add more blocks:
     Uncaught AssertionError: expected 'invalid Difficulty' to not exist
      at eth.vm._blockchain.putBlock (test/vm-with-real-chain-1st-1000.spec.js:49:29)
      at node_modules/ethereumjs-blockchain/index.js:114:5
      at node_modules/ethereumjs-blockchain/node_modules/async/lib/async.js:726:13
```

To try it out, git clone, npm i, take of the `skip` from this test and npm run test. Note that putting the genesis block runs fine though :)

**Pull request merged**